samples/HttpServer_FirmwareUpload/app/application.cpp

```diff
@@ -3,10 +3,120 @@
 #include <Data/Stream/LimitedMemoryStream.h>
 #include <MultipartParser/HttpMultipartResource.h>
 #include <Data/Stream/RbootOutputStream.h>
+#include <sodium/crypto_sign.h>
+#include "FirmwareVerificationKey.h"
 
 HttpServer server;
 String lastModified;
 
+class SignedRbootOutputStream : public RbootOutputStream
+{
+	const uint8_t* const verificationKey_;
+
+	struct {
+		uint32_t magic;
+		uint32_t loadAddress;
+		uint8_t signature[crypto_sign_BYTES];
+	} header_;
+
+	uint8_t* headerPtr_;
+	size_t missingHeaderBytes_;
+	static const uint32_t HEADER_MAGIC_EXPECTED = 0xf01af02a;
+
+	bool ok_;
+	bool closeCalled_ = false;
+	uint32_t startAddress_;
+
+	crypto_sign_state verifierState_;
+	String errorMessage_;
+
+	void setError(const char* message)
+	{
+		ok_ = false;
+		errorMessage_ = message;
+		Serial.println(errorMessage_);
+	}
+
+public:
+	SignedRbootOutputStream(int32_t startAddress, size_t maxLength, const uint8_t* verificationKey)
+		: RbootOutputStream(startAddress, maxLength), verificationKey_(verificationKey),
+		  headerPtr_(reinterpret_cast<uint8_t*>(&header_)), missingHeaderBytes_(sizeof(header_)), ok_(true),
+		  startAddress_(startAddress)
+	{
+		crypto_sign_init(&verifierState_);
+	}
+
+	bool ok() const
+	{
+		return ok_;
+	}
+	const String& errorMessage() const
+	{
+		return errorMessage_;
+	}
+
+	size_t write(const uint8_t* data, size_t size)
+	{
+		size_t consumed = 0;
+		if(ok_) {
+			if(missingHeaderBytes_ > 0) {
+				const size_t chunkSize = std::min(size, missingHeaderBytes_);
+				memcpy(headerPtr_, data, chunkSize);
+				headerPtr_ += chunkSize;
+				missingHeaderBytes_ -= chunkSize;
+				data += chunkSize;
+				size -= chunkSize;
+				consumed += chunkSize;
+				if(missingHeaderBytes_ == 0) {
+					Serial.printf("Receive image for load address 0x%08X, slot starts at 0x%08X\n", header_.loadAddress,
```

Comment:
Please, use `debug_d` for useful debug messages. For example the line above should become:
```
debug_d(("Receive image for load address 0x%08X, slot starts at 0x%08X\n", header_.loadAddress,startAddress_)
```

Reply:
Fixed.
However, since this is an example, I thought it would be beneficial to show some output during the upgrade, regardless of the debug settings.